pkg/ingester/user_metrics_metadata_test.go

```diff
@@ -0,0 +1,120 @@
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "qux"},
+			},
+		},
+		"should fail when metadata per user limit reached": {
+			maxMetadataPerUser: 1,
+			inputMetadata: []input{
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"}, errorExpected: false},
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"}, errorExpected: false},
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"}, errorExpected: true},
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "qux"}, errorExpected: true},
+			},
+			expectedMetadata: []*mimirpb.MetricMetadata{
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"},
+			},
+		},
+		"should fail when metadata per metric limit reached": {
+			maxMetadataPerMetric: 1,
+			inputMetadata: []input{
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"}, errorExpected: false},
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"}, errorExpected: true},
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"}, errorExpected: false},
+				{meta: mimirpb.MetricMetadata{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "qux"}, errorExpected: true},
+			},
+			expectedMetadata: []*mimirpb.MetricMetadata{
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"},
+			},
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			// Mock the ring
+			ring := &ringCountMock{}
+			ring.On("HealthyInstancesCount").Return(1)
+			ring.On("ZonesCount").Return(1)
+
+			// Mock limiter
+			limits, err := validation.NewOverrides(validation.Limits{
+				MaxGlobalMetricsWithMetadataPerUser: testData.maxMetadataPerUser,
+				MaxGlobalMetadataPerMetric:          testData.maxMetadataPerMetric,
+			}, nil)
+			require.NoError(t, err)
+			limiter := NewLimiter(limits, ring, 1, false)
+
+			// Mock metrics
+			metrics := newIngesterMetrics(
+				prometheus.NewPedanticRegistry(),
+				true,
+				func() *InstanceLimits { return defaultInstanceLimits },
+				nil,
+				nil,
+			)
+
+			mm := newMetadataMap(limiter, metrics, "test")
+
+			// Attempt to add all metadata
+			for _, i := range testData.inputMetadata {
+				err := mm.add(i.meta.MetricFamilyName, &i.meta)
+
+				if i.errorExpected {
+					require.Error(t, err)
```

Comment:
since you're testing two possible reasons for errors, it might be worth it to also check whether you're getting the correct error back

Reply:
addressed in https://github.com/grafana/mimir/pull/2184/commits/779ea09787f7209f290a6b4ed171991a348f3d97. Thank you!